src/compiler/checker.ts

```diff
@@ -44826,7 +44826,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let meaning = SymbolFlags.None;
             // In an interface or class, we're definitely interested in a type.
             if (name.parent.kind === SyntaxKind.ExpressionWithTypeArguments) {

```

Comment:
This statement here still effectively states that we're doing the buggy behavior (i.e. assuming that `ExpressionWithTypeArguments` only appears in heritage clauses, what became false when instantiation expressions were added); maybe we should remove this comment or rephrase it?

Reply:
Yep, I had no idea what heritage clauses had to do with the bug until Jake pointed this out.

Reply:
E.g. above the new code below, something like:

> An InstantiationExpression may appear, in type space (`interface Foo<T> extends Bar<T>`), value space (`return foo<T>`), or both (`class Foo<T> extends Bar<T>`); ensure the meaning matches.